### ISSUE TITLE
Invalid assertion in conch.telnet option negotiation.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4231]
|<img alt="spotter's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| spotter reported|
|-|-|
|Trac ID|trac#4231|
|Type|defect|
|Created|2010-01-21 11:23:53Z|
|Branch|https://github.com/twisted/twisted/tree/telnet-option-assert-4231|

**Short:** Assertion in dont_no_true breaks after WILL-DONT exchange.

**How to trigger:**

1. Offer opt by issuing will(opt); You get opt state of: us=no*, him=no;

2. Wait for response, where two things may happen:

    2.1 Response is do(opt), do_no_true is called, state is now: us=yes, him=no; and we live to see another data chunk. ('''CORRECT''')

    2.2 Response is dont(opt), dont_no_true is called, we land in never-never land and die due to assertion error. ('''BAD''')

**Possible solution:**

Wouldn't it be better to just end negotiation if we get dont(opt) here? Currently my solution is to overload in my transport:

```python
    def dont_no_true(self, state, option):
        state.us.negotiating = False
```
```python
    dontMap = Telnet.dontMap
    dontMap[('no', True)] = dont_no_true
```

I did not took time to look at other assertions in opt negotiations, but is it really the right way to kill object due to something like this? Wouldn't error log message be sufficient?

(Version: Twisted 9.0.0)

Attachments:

* [dont_wont.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/04c/04c5be6ef82e5c1d2b2339d4d2bb57efab89903a/1573d1751d8e8f49ccd076fbef8ddac094b17b15.diff) (4025 bytes) - added by spotter on 2010-01-24 23:17:16Z - Patch for test_telnet.py and telnet.py

<details><summary>Searchable metadata</summary>

```
trac-id__4231 4231
type__defect defect
reporter__spotter spotter
priority__normal normal
milestone__ 
branch__branches_telnet_option_assert_4231 branches/telnet-option-assert-4231
branch_author__exarkun__spotter exarkun, spotter
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__conch_telnet_option_negotiation_assertion conch telnet option negotiation assertion
time__1264073033000000 1264073033000000
changetime__1286494079000000 1286494079000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="kench's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|kench commented|
|-|-|

Shouldn't it also trigger the deferred to let the caller of **ITelnetTransport.will** know that the request was vetoed?  The doc string for **ITelnetTransport.will** suggests that it should be triggering the callback with **False**, though the doc string for **ITelnetTransport.do** says about the same thing and **Telnet.wont_no_true** is actually triggering the errback with an **OptionRefused(option)**.

```
    def dont_no_true(self, state, option):
        # Peer refused our request to enable an option.
        state.us.negotiating = False
        d = state.us.onResult
        state.us.onResult = None
        d.callback(False)
         -- or maybe --
        d.errback(OptionRefused(option))
```

**Comment by twisted-trac** [comment]
|<img alt="spotter's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|spotter commented|
|-|-|

I would actually go for anything that is not the assert, maybe slightly in favour of errback, since it would be then symmetric to do(opt). Any of these two would work for me as well, since I could just handle it along with option enabling callback. 

In my use case nothing needs to be done if negotiation rolls that way, so I kept it simple and just finalised the process.

**Comment by twisted-trac** [comment]
|<img alt="spotter's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|spotter commented|
|-|-|

Replying to [kench](#note_1):
> Shouldn't it also trigger the deferred to let the caller of **ITelnetTransport.will** know that the request was vetoed?  The doc string for **ITelnetTransport.will** suggests that it should be triggering the callback with **False**, though the doc string for **ITelnetTransport.do** says about the same thing and **Telnet.wont_no_true** is actually triggering the errback with an **OptionRefused(option)**.

I've actually spent time to review the docstrings of ITelnetTransport and it should indeed run callback(False) in both cases, so OptionRefused(Opt) is not correct either. I've modified my Transport to reflect this:

```python
    # overloading Telne.dont_no_true 
    def dont_no_true(self, state, option):
        '''Proposed option was refused.'''
        # Patching twisted bug to ensure promised behaviour.
        state.us.negotiating = False
        d = state.us.onResult
        state.us.onResult = None
        d.callback(False)
    # patching Telnet.dontMap
    dontMap = Telnet.dontMap
    dontMap[('no', True)] = dont_no_true
    # overloading Telnet.wont_no_true
    def wont_no_true(self, state, option):
        '''Peer refused to enable an option in response to our request.'''
        state.him.negotiating = False
        d = state.him.onResult
        state.him.onResult = None
        d.callback(False) # Patching twisted bug to ensure promised behaviour.
    # patching Telnet.wontMap
    wontMap = Telnet.wontMap
    wontMap[('no', True)] = wont_no_true
```

Would that be the correct way of solving this problem?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Thanks for investigating this further.  I haven't had a chance to look closely at this issue yet, but I know a couple things will make it easier to answer that last question:

  1. Supply proposed changes as patches in unified diff format ('svn diff' generates these by default)
  1. Start off with a unit test for behavior in question, including a docstring which explains what behavior is expected.  Having a unit test that fails without the proposed change and passes with it is a big help to others for understanding the proposed change.

**Comment by twisted-trac** [comment]
|<img alt="kench's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|kench commented|
|-|-|

I'm no authority on what is or isn't correct -- just an interested party.  

Personally, I'd be happy with either behavior -- d.callback(False) or 
d.errback(OptionRefused(option)).  I just think it ought to be consistent and correctly documented.  

Since there may already be users in the wild expecting the OptionRefused behavior for the do/wont case, that may be the less disruptive direction to go.  

From a more philosophical line of reasoning, refusing an option isn't really an "exception" as such -- just part of negotiating.  That would suggest that returning False to the callback would be more "correct".  

There -- is that wishy-washy enough?  I oughta go into politics.

**Comment by twisted-trac** [comment]
|<img alt="spotter's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|spotter commented|
|-|-|

Replying to [exarkun](#note_4):
> Thanks for investigating this further.  I haven't had a chance to look closely at this issue yet, but I know a couple things will make it easier to answer that last question:
>
>
>   1. Supply proposed changes as patches in unified diff format ('svn diff' generates these by default)


Attached. Added one test, modified the other.


>   2. Start off with a unit test for behavior in question, including a docstring which explains what behavior is expected.  Having a unit test that fails without the proposed change and passes with it is a big help to others for understanding the proposed change.


Added comments to test methods, pointing to documentation in ITelnetProtocol, the same that was quoted by kench in his first comment. I've modified behaviour of two methods in Telnet to match said docstrings. As kench noted: these changes may break backward compatibility, although they actually make Telnet working in accordance to ITelnetProtocol promise. As far as I could test new test_telnet.py seems to be OK and I have not seen any regressions in test_manhole.py.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [30103]) Branching to 'telnet-option-assert-4231'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [30104]) Apply dont_wont.diff

Refs [#4231](https://github.com/twisted/twisted/issues/4231)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [30105]) Add docstrings to _OptionState and _Perspective

refs [#4231](https://github.com/twisted/twisted/issues/4231)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [30106]) Rename the test method to conform to the naming standard; give it a docstring

refs [#4231](https://github.com/twisted/twisted/issues/4231)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks for the patch.  I messed around with docs for a while to convince myself that the change is really necessary, and eventually did.  Then I moved all the doc changes to [#4677](https://github.com/twisted/twisted/issues/4677) since I also reformatted stuff and that would have just cluttered up this diff.

As far as the `False` vs `OptionRefused`, since `OptionRefused` is what was implemented already, I think changing the documentation is better.  This reduces the likelihood of breaking code using these APIs.  It's less likely that a documentation change will break things.

I went ahead and adjusted the branch to reflect this.  That meant dropping the changes to the existing test and `wont_no_true` and changing the `callback(False)` in `dont_no_true` to an `errback(OptionRefused(...))`, and changing the tests to match.

This fix makes sense to me now, so I'll probably merge it (after I see the [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/telnet-option-assert-4231)).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [30118]) Merge telnet-option-assert-4231

Author: spotter, exarkun
Reviewer: exarkun
Fixes: [#4231](https://github.com/twisted/twisted/issues/4231)

Handle a DONT response to a WILL request by firing the waiting Deferred
with an `OptionRefused` exception instead of triggering an assert statement.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
